src/Soil/Soil.jl

```diff
@@ -165,30 +192,208 @@ function dss!(
 
+A simple concrete type of boundary condition, which enforces a
+constant normal state at either the top and bottom of the domain.
 """
-   FluxBC{FT} <: AbstractSoilBoundaryConditions{FT}
+struct StateBC{FT} <: AbstractBC{FT}
+    bc::FT
+end
 
-A simple concrete type of boundary condition, which enforces
-constant normal fluxes at the top and bottom of the domain.
 """
-struct FluxBC{FT} <: AbstractSoilBoundaryConditions{FT}
-    top_flux_bc::FT
-    bot_flux_bc::FT
+   FluxBC{FT} <: AbstractBC{FT}
+
+A simple concrete type of boundary condition, which enforces a
+constant normal flux at either the top or bottom of the domain.
+"""
+struct FluxBC{FT} <: AbstractBC{FT}
+    bc::FT
 end
 
 """
-    boundary_fluxes(bc::FluxBC, _...)
+    TopBoundary{} <: AbstractBoundary{}
 
-A function which returns the correct boundary flux
-given the boundary condition type `FluxBC`.
+A simple object which should be passed into a function to
+indicate that we are considering the top boundary of the soil.
+"""
+struct TopBoundary <: AbstractBoundary end
+
+"""
+    BotBoundary{} <: AbstractBoundary{}
+
+A simple object which should be passed into a function to
+indicate that we are considering the top boundary of the soil.
+"""
+struct BotBoundary <: AbstractBoundary end
 
-This is a trivial example, but a more complex one would be e.g.
-Dirichlet conditions on the state, which then must be converted into
-a flux before being applied as a boundary condition.
 """
-function boundary_fluxes(bc::FluxBC, _...)
-    return bc.top_flux_bc, bc.bot_flux_bc
+    RREBoundaryConditions{FT, T_top <: AbstractBC{FT}, T_bot <: AbstractBC{FT}} 
+        <: AbstractSoilBoundaryConditions{FT}
+
+A wrapper for soil boundary conditions in the water-only case.
+This can contain top and bottom boundary conditions of any types
+(Flux, State, etc).
+"""
+struct RREBoundaryConditions{
+    FT,
+    T_top <: AbstractBC{FT},
+    T_bot <: AbstractBC{FT},
+} <: AbstractSoilBoundaryConditions{FT}
+    bc_top::T_top
+    bc_bot::T_bot
+end
+
+function RREBoundaryConditions{FT}(
+    bc_top::AbstractBC{FT},
+    bc_bot::AbstractBC{FT},
+) where {FT}
+    return RREBoundaryConditions{FT, typeof(bc_top), typeof(bc_bot)}(
+        bc_top,
+        bc_bot,
+    )
+end
+
+"""
+    RREHeatBoundaryConditions{FT, T_rre_top <: AbstractBC{FT}, T_rre_bot <: AbstractBC{FT},
+    T_heat_top <: AbstractBC{FT}, T_heat_bot <: AbstractBC{FT}} 
+    <: AbstractSoilBoundaryConditions{FT}
+
+A wrapper for soil boundary conditions in the water-and-heat case.
+This can contain top and bottom boundary conditions of any types
+(Flux, State, etc).
+"""
+struct RREHeatBoundaryConditions{
+    FT,
+    T_rre_top <: AbstractBC{FT},
+    T_heat_top <: AbstractBC{FT},
+    T_rre_bot <: AbstractBC{FT},
+    T_heat_bot <: AbstractBC{FT},
+} <: AbstractSoilBoundaryConditions{FT}
+    bc_top::Tuple{T_rre_top, T_heat_top}
+    bc_bot::Tuple{T_rre_bot, T_heat_bot}
+end
+
+function RREHeatBoundaryConditions{FT}(
```

Comment:
do we need these type containers? Or could we use named Tuples? (heat = (top = , bot = ), water = (top = , bot = ))